File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        ref = {inorder[i] : i for i in range(len(inorder))}

        def dfs(preorder, inorder): 
            nonlocal ref 
            root = TreeNode(preorder[0])
            index = ref[preorder[0]]
            if index == 0 and index == len(preorder) - 1:
                root.left = None
                root.right = None 
            elif index == 0:
                root.left = None 
                root.right = self.buildTree(preorder[index + 1:], inorder[index + 1 :])
            elif index == len(preorder) - 1:
                root.left = self.buildTree(preorder[1 : index + 1], inorder[:index])
                root.right = None
            else:
                root.left = self.buildTree(preorder[1 : index + 1], inorder[:index])
                root.right = self.buildTree(preorder[index + 1:], inorder[index + 1 :])
            return root
        
        return dfs(preorder, inorder)

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {

    int target;
    int counter = 0;

    public int kthSmallest(TreeNode root, int k) {
        dfs(root, k);
        return target;
        
    }

    void dfs(TreeNode node, int k){
        if (node == null) {
            return;
        }

        dfs(node.left, k);
        if (counter > k) {
            return;
        }
        counter += 1;
        if (counter == k){
            target = node.val;
        }
        dfs(node.right, k);

    }
}
